proposals/char-kind-proposal.md

```diff
@@ -0,0 +1,173 @@
+
+```haskell
+type family CharToNat (a :: Char) :: Nat
+
+type family NatToChar (a :: Nat) :: Char
+```
+
+A type-level analogue of the function `generalCategory` from `Data.Kind`.
+
+```haskell
+type family GeneralCharCategory (a :: Char) :: GeneralCategory
+```
+
+
+The second group of type families consists of built-in unary predicates. All of them are based on their corresponding term-level analogues from `Data.Char`. The precise list is the following one:
+
+```haskell
+type family IsAlpha (a :: Char) :: Bool
+
+type family IsAlphaNum (a :: Char) :: Bool
+
+type family IsControl (a :: Char) :: Bool
+
+type family IsPrint (a :: Char) :: Bool
+
+type family IsUpper (a :: Char) :: Bool
+
+type family IsLower (a :: Char) :: Bool
+
+type family IsSpace (a :: Char) :: Bool
+
+type family IsDigit (a :: Char) :: Bool
+
+type family IsOctDigit (a :: Char) :: Bool
+
+type family IsHexDigit (a :: Char) :: Bool
+
+type family IsLetter (a :: Char) :: Bool
+```
+
+We also provide several type-level predicates implemented via the `GeneralCharCategory` type family.
+
+```haskell
+type IsMark a = IsMarkCategory (GeneralCharCategory a)
+
+type IsNumber a = IsNumberCategory (GeneralCharCategory a)
+
+type IsPunctuation a = IsPunctuationCategory (GeneralCharCategory a)
+
+type IsSymbol a = IsSymbolCategory (GeneralCharCategory a)
+
+type IsSeparator a = IsSeparatorCategory (GeneralCharCategory a)
+```
+
+Built-in type families we described above are supported with the corresponding definitions and functions in the modules `compiler/GHC/Builtin/Names.hs`, `compiler/GHC/Builtin/Types.hs`, and `compiler/GHC/Builtin/Types/Literals.hs`. In addition to type families, our patch contain the following updates:
+
+1. parsing the 'x' syntax
+2. type-checking `'x' :: Char`
+3. type-checking `Refl :: 'x' :~: 'x'`
+4. Typeable / TypeRep support
+5. template-haskell support
+6. Haddock related updates
+7. tests
+
+
+## Examples
+
+The following code snippets are well-defined with our patch:
+
+The example of type level character comparison:
+```haskell
+f1 :: CmpChar 'x' 'x' :~: EQ
+f1 = Refl
+
+f2 :: CmpChar 'x' 'y' :~: LT
+f2 = Refl
+```
+
+The following examples are related to type families based on the corresponding functions from `Data.Char`:
+```haskell
+testIsLower :: '(IsLower 'x', IsLower 'X') :~: '(True, False)
+testIsLower = Refl
+
+testIsUpper :: '(IsUpper 'x', IsUpper 'X') :~: '(False, True)
+testIsUpper = Refl
+
+testIsAlpha :: '[IsAlpha 'X', IsAlpha 'x', IsAlpha '6', IsAlpha '/']
+  :~: '[True, True, False, False]
+testIsAlpha = Refl
+
+testIsAlphaNum :: '[IsAlphaNum ',', IsAlphaNum 'x', IsAlphaNum '6', IsAlphaNum '/']
+  :~: '[False, True, True, False]
+testIsAlphaNum = Refl
+```
+
+## Effect and Interactions
+
+Our patch allows one to analyse type-level strings more precisely having promoted chars and helpers that almost entirely cover term-level functions from `Data.Char`. At the moment, we haven't identified any contentious interactions yet.
```

Comment:
I'm struggling to understand the grammatical structure of the first sentence. Even if it's valid, I think the "having ..." clause is way too fancy. Also, what do you mean by saying that type-level strings can be analyzed "more precisely"? What sort of imprecision exists today?